### ISSUE TITLE
Nginx + Gzip Static + Modpagespeed 

### ISSUE THREAD

**Issue by oschaaf** [pagespeed/ngx_pagespeed#1053]
From https://groups.google.com/d/msgid/mod-pagespeed-discuss/ac4d7690-a361-4964-85e9-4e6aafd12d28%40googlegroups.com?utm_medium=email&utm_source=footer

I am running Nginx 1.9.7 with modpagespeed (see version info below).  Modpagespeed is working correctly.  However, if I set 'gzip_static always;', static .gz files are not served when they are present.  If I set 'pagespeed off;' then the gzip_static takes effect and Nginx correctly serves the compressed files.  I am using the test files from https://github.com/mathiasbynens/nginx-zopfli-test as a reference point.  Is this a known issue with modpagespeed?

Nginx version info:
nginx version: nginx/1.9.7
built by gcc 4.9.2 (Debian 4.9.2-10)
built with OpenSSL 1.0.2d 9 Jul 2015
TLS SNI support enabled

configure arguments: 

--prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-http_ssl_module --with-http_realip_module --with-http_addition_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_random_index_module --with-http_secure_link_module --with-http_stub_status_module --with-http_auth_request_module --add-module=/opt/rebuildnginx/nginx-1.9.7/debian/modules/ngx_pagespeed --with-pcre=/opt/rebuildnginx/nginx-1.9.7/debian/modules/pcre --with-pcre-jit --with-openssl=/opt/rebuildnginx/nginx-1.9.7/debian/modules/openssl --with-zlib=/opt/rebuildnginx/nginx-1.9.7/debian/modules/zlib --with-threads --with-stream --with-stream_ssl_module --with-mail --with-mail_ssl_module --with-file-aio --with-http_v2_module --with-cc-opt='-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' --with-ld-opt=-Wl,-z,relro --with-ipv6

**Comment by LukeMauldin** [comment]
Setting "pagespeed inPlaceResourceOptimization off;" allowed the gzip_static directive to work.  So this issue only occurs when inPlaceResourceOptimization is on.

**Comment by morlovich** [comment]
Are affected files all css and js, or was anything else affected?

**Comment by jmarantz** [comment]
"Setting the option "pagespeed inPlaceResourceOptimization off;" allowed the gzip_static directive to work.  What are the disadvantages to setting this to off?"

If you set this off, only resources whose URLs found in CSS and HTML markup can be rewritten, by changing their URLs.  However I want to point out that I think this doesn't really fix the problem.

If I understand correctly, you have zopfli-compressed your CSS & JS files to make them smaller.  PageSpeed has uncompressed those files, miinified them, and then let nginx gzip them using default settings, which are not as small as Zopfli.  It might be the case that PageSpeed made the uncompressed file 1% smaller but then lost 10% by compressing it less aggressively.  Obviously that's suboptimal :)

Whether that happens via InPlaceResourceOptimization or by rewriting URLs doesn't matter too much IMO.


"When will the new version be released?"  Soon, hopefully this week.  This one has been a long time coming.

"Will it enable zopfli compression?"" No, but it will at least statically compress the PageSpeed optimized output with gzip -9.  In a future release we are looking at incorporating Zopfli or Brotli.


I am going to rename this bug to reflect what I think is actually a bug:

**Comment by jmarantz** [comment]
Title was: Nginx + Gzip Static + Modpagespeed

**Comment by LukeMauldin** [comment]
Thanks.  I only tested .js files.  Did not test .css files.

**Comment by LukeMauldin** [comment]
What tagged version should I be looking for that includes these fixes?

**Comment by jeffkaufman** [comment]
@LukeMauldin 1.10.33.0 will have these fixes, when it's released.

**Comment by nikolay** [comment]
@jeffkaufman did it fix it?

**Comment by jeffkaufman** [comment]
@nikolay @LukeMauldin I'm sorry, I misread the history of this issue when I commented on 11/30.  We haven't fixed this yet.  I've added it to the [prioritization page](https://github.com/pagespeed/mod_pagespeed/wiki/Work-Prioritization) but there are quite a few things ahead of it.  Sorry for the false hope!

**Comment by jmarantz** [comment]
IIUC this can be fixed only by switching to Zopfli for our gzip compressor. Is that right?

**Comment by jeffkaufman** [comment]
No, we should be able to just not recompress existing gzipped resources, or
only recompress them if we make them smaller.

**Comment by jmarantz** [comment]
Jeff -- the flow I'm thinking about is:

   1. origin JS file
   2. manual aggressive compression (e.g. using zopfli)
   3. pagespeed unzips the origin JS, minifies it, and removes 1% of the
   bytes in cleartext
   4. pagespeed re-gzips the minified file with "gzip -9"

At which point in the flow should we say "hey we didn't make it smaller
than the origin"?  We can definitely do this if we properly record
x-original-content-length when we unzip it, and check that when we re-gzip
it.

Also note that while rewrite_javascript is a pure minification, the CSS
rewriter actually makes other transformations that might make the CSS
slightly bigger but better (e.g. flattening, rewriting embedded images), so
a pure size comparison in the system core may not be ideal.

So we need to know at the time we re-gzip a file that in the end it didn't
help and we might as well have just served the origin bytes unaltered.

WDYT?
